File: backend/src/controllers/orderController.js
const { pool } = require('../config/database');
const { generateOrderId } = require('../utils/validation');

// Create Order
const createOrder = async (req, res) => {
  try {
    const { amount, currency = 'INR', receipt, notes } = req.body;
    const merchantId = req.merchant.id;

    // Validate amount
    if (!amount || amount < 100) {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'amount must be at least 100'
        }
      });
    }

    // Generate unique order ID
    let orderId = generateOrderId();
    let isUnique = false;
    
    while (!isUnique) {
      const existingOrder = await pool.query('SELECT id FROM orders WHERE id = $1', [orderId]);
      if (existingOrder.rows.length === 0) {
        isUnique = true;
      } else {
        orderId = generateOrderId();
      }
    }

    // Create order
    const result = await pool.query(
      `INSERT INTO orders (id, merchant_id, amount, currency, receipt, notes, status, created_at, updated_at)
       VALUES ($1, $2, $3, $4, $5, $6, 'created', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
       RETURNING *`,
      [orderId, merchantId, amount, currency, receipt, notes ? JSON.stringify(notes) : null]
    );

    const order = result.rows[0];

    res.status(201).json({
      id: order.id,
      merchant_id: order.merchant_id,
      amount: order.amount,
      currency: order.currency,
      receipt: order.receipt,
      notes: order.notes || {},
      status: order.status,
      created_at: order.created_at.toISOString()
    });
  } catch (error) {
    console.error('Error creating order:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get Order
const getOrder = async (req, res) => {
  try {
    const { order_id } = req.params;
    const merchantId = req.merchant.id;

    const result = await pool.query(
      'SELECT * FROM orders WHERE id = $1 AND merchant_id = $2',
      [order_id, merchantId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: {
          code: 'NOT_FOUND_ERROR',
          description: 'Order not found'
        }
      });
    }

    const order = result.rows[0];

    res.status(200).json({
      id: order.id,
      merchant_id: order.merchant_id,
      amount: order.amount,
      currency: order.currency,
      receipt: order.receipt,
      notes: order.notes || {},
      status: order.status,
      created_at: order.created_at.toISOString(),
      updated_at: order.updated_at.toISOString()
    });
  } catch (error) {
    console.error('Error getting order:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get Order (Public - for checkout page)
const getOrderPublic = async (req, res) => {
  try {
    const { order_id } = req.params;

    const result = await pool.query(
      'SELECT id, amount, currency, status FROM orders WHERE id = $1',
      [order_id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: {
          code: 'NOT_FOUND_ERROR',
          description: 'Order not found'
        }
      });
    }

    const order = result.rows[0];

    res.status(200).json({
      id: order.id,
      amount: order.amount,
      currency: order.currency,
      status: order.status
    });
  } catch (error) {
    console.error('Error getting order:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

module.exports = { createOrder, getOrder, getOrderPublic };

File: backend/src/main/java/com/paymentgateway/controller/OrderController.java
package com.paymentgateway.controller;

import com.paymentgateway.dto.ErrorResponse;
import com.paymentgateway.dto.OrderRequest;
import com.paymentgateway.dto.OrderResponse;
import com.paymentgateway.entity.Merchant;
import com.paymentgateway.entity.Order;
import com.paymentgateway.repository.OrderRepository;
import com.paymentgateway.repository.MerchantRepository;
import com.paymentgateway.util.ValidationUtil;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Optional;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

@RestController
@RequestMapping("/api/v1")
public class OrderController {
    
    @Autowired
    private OrderRepository orderRepository;
    
    @Autowired
    private MerchantRepository merchantRepository;
    
    /**
     * POST /api/v1/orders
     * Create an order
     */
    @PostMapping("/orders")
    public ResponseEntity<?> createOrder(@RequestBody OrderRequest request, HttpServletRequest httpRequest) {
        try {
            Merchant merchant = (Merchant) httpRequest.getAttribute("merchant");
            if (merchant == null) {
                return ResponseEntity.status(401).body(
                    new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials")
                );
            }
            
            // Validate amount
            if (request.amount == null || request.amount < 100) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("BAD_REQUEST_ERROR", "Amount must be at least 100 paise")
                );
            }
            
            // Create order
            String orderId = ValidationUtil.generateOrderId();
            Order order = new Order(orderId, merchant, request.amount);
            
            if (request.currency != null) {
                order.setCurrency(request.currency);
            }
            if (request.receipt != null) {
                order.setReceipt(request.receipt);
            }
            if (request.notes != null) {
                ObjectMapper mapper = new ObjectMapper();
                JsonNode notesNode = mapper.valueToTree(request.notes);
                order.setNotes(notesNode);
            }
            
            orderRepository.save(order);
            
            // Build response
            OrderResponse response = new OrderResponse();
            response.id = order.getId();
            response.amount = order.getAmount();
            response.currency = order.getCurrency();
            response.receipt = order.getReceipt();
            response.status = order.getStatus();
            response.created_at = order.getCreatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
            
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (Exception e) {
            System.err.println("Error creating order: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * GET /api/v1/orders/{id}
     * Get order details (merchant only)
     */
    @GetMapping("/orders/{id}")
    public ResponseEntity<?> getOrder(@PathVariable String id, HttpServletRequest httpRequest) {
        try {
            Merchant merchant = (Merchant) httpRequest.getAttribute("merchant");
            if (merchant == null) {
                return ResponseEntity.status(401).body(
                    new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials")
                );
            }
            
            Optional<Order> order = orderRepository.findByIdAndMerchant(id, merchant);
            if (order.isEmpty()) {
                return ResponseEntity.status(404).body(
                    new ErrorResponse("NOT_FOUND_ERROR", "Order not found")
                );
            }
            
            Order o = order.get();
            OrderResponse response = new OrderResponse();
            response.id = o.getId();
            response.amount = o.getAmount();
            response.currency = o.getCurrency();
            response.receipt = o.getReceipt();
            response.status = o.getStatus();
            response.created_at = o.getCreatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            System.err.println("Error getting order: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * GET /api/v1/orders/{id}/public
     * Get order details (public - for checkout page)
     */
    @GetMapping("/orders/{id}/public")
    public ResponseEntity<?> getOrderPublic(@PathVariable String id) {
        try {
            Optional<Order> order = orderRepository.findById(id);
            if (order.isEmpty()) {
                return ResponseEntity.status(404).body(
                    new ErrorResponse("NOT_FOUND_ERROR", "Order not found")
                );
            }
            
            Order o = order.get();
            OrderResponse response = new OrderResponse();
            response.id = o.getId();
            response.amount = o.getAmount();
            response.currency = o.getCurrency();
            response.receipt = o.getReceipt();
            response.status = o.getStatus();
            response.created_at = o.getCreatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
            
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            System.err.println("Error getting order: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
}
